File: 2015/day_6/main.py
import time

if __name__ == "__main__":
    with open("input.txt", "r") as f:
        input = f.read()

    start_time = time.time()
    input = input.split("\n")

    # create the grid by double list comprehension
    on_grid = [[False for i in range(1000)] for i in range(1000)]
    brightness_grid = [[0 for i in range(1000)] for i in range(1000)]
    
    for line in input:
        # split input into words and get necessary information
        line = line.split(" ")
        if line[0] == "toggle":
            start = line[1].split(",")
            command = "toggle"
        else:
            start = line[2].split(",")
            command = line[1]
        
        end = line[-1].split(",")
        x1 = int(start[0])
        y1 = int(start[1])
        x2 = int(end[0])
        y2 = int(end[1])

        # iterate over the grid and change values
        while x1 <= x2:
            j = y1
            while j <= y2:
                if command == "toggle":
                    val = on_grid[x1][j]
                    on_grid[x1][j] = not val
                    brightness_grid[x1][j] += 2
                elif command == "on":
                    on_grid[x1][j] = True
                    brightness_grid[x1][j] += 1
                elif command == "off":
                    on_grid[x1][j] = False
                    brightness_grid[x1][j] -= 1 if brightness_grid[x1][j] > 0 else 0
                j += 1
            x1 +=1

    count_on_lights = sum([x.count(True) for x in on_grid])
    count_on_brightness = sum([sum(x) for x in brightness_grid])

    end = time.time()
    
    print(f"{count_on_lights} lights are on according to santas plan")
    print(f"{count_on_brightness} is the brightness on according to santas plan")
    print(f"Needed {end - start_time} seconds")

File: 2015/day_6/main.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include "line.h"

typedef struct grid{
    int values[1000][1000];
}grid;

grid* createArray() {
    grid *newArray = malloc(sizeof(grid));

    for (int i=0; i < 1000; i++) {
        for (int j=0; j < 1000; j++) {
            newArray->values[i][j] = 0;
        }
    }

    return newArray;
}

int main() {
    FILE *fp = fopen("input.txt", "r");

    if (!fp) {
        printf("No File found!\n");
        return 1;
    }

    clock_t start_time = clock();

    grid* onGrid = createArray();
    grid* brightnessGrid = createArray();

    char *line = NULL;
    size_t len = 0;
    ssize_t read;

    char *whiteSpacePtr;
    char * commaPtr;
    char buffer[4];

    int onLights = 0;
    int brightness = 0;

    while ((read = getLine(&line, &len, fp)) != -1 ) {
        // get the command
        whiteSpacePtr = strchr(line, ' ');
        int length = whiteSpacePtr - line;

        // -1 turn off
        // 1 turn on
        // 2 toggle
        int command = 0;

        if (length == 4) {
            whiteSpacePtr = strchr(whiteSpacePtr + 1, ' ');
            length = whiteSpacePtr - line;
            if (length == 8) {
                command = -1;
            }
            else {
                command = 1;
            }
        }
        else { 
            command = 2;
        }
        
        // use incremental whitespace search to identify the 
        // integers for next part
        commaPtr = strchr(line, ',');

        strncpy(buffer, whiteSpacePtr, commaPtr - whiteSpacePtr);
        int x1 = atoi(buffer);
       
        memset(buffer, 0, strlen(buffer));
        whiteSpacePtr = strchr(commaPtr, ' ');
        strncpy(buffer, commaPtr + 1, whiteSpacePtr - commaPtr);
        int y1 = atoi(buffer);

        memset(buffer, 0, strlen(buffer));
        commaPtr = strrchr(line, ',');
        strncpy(buffer, commaPtr +1, strlen(commaPtr+1));
        int y2 = atoi(buffer);

        memset(buffer, 0, strlen(buffer));
        whiteSpacePtr = strrchr(line, ' ');
        strncpy(buffer, whiteSpacePtr + 1, commaPtr - whiteSpacePtr - 1);
        int x2 = atoi(buffer);

        memset(buffer,0,strlen(buffer));

        // iterate over necessary parts of grid
        while (x1 <= x2) {
            int j = y1;
            while (j <= y2) {
                switch (command) {
                    case -1:
                        onGrid->values[x1][j] = 0;
                        int currentValue = brightnessGrid->values[x1][j];
                        brightnessGrid->values[x1][j] = (currentValue > 0) ? (--currentValue) : 0;
                        break;
                    case 1:
                        onGrid->values[x1][j] = 1;
                        brightnessGrid->values[x1][j] +=1;
                        break;
                    case 2: 
                        onGrid->values[x1][j] = (onGrid->values[x1][j] == 0) ? 1 : 0;
                        brightnessGrid->values[x1][j] += 2;
                        break;
                    default:
                        break;
                }
                j++;
            }
            x1++;
        }
    }

    // iterate over whole grid and count
    for (int i=0; i < 1000; i++) {
        for (int j = 0; j < 1000; j++) {
            onLights += onGrid->values[i][j];
            brightness += brightnessGrid->values[i][j];
        }
    }

    clock_t end_time = clock();

    free(fp);
    free(line);
    free(whiteSpacePtr);
    free(commaPtr);
    free(onGrid);
    free(brightnessGrid);

    printf("%d lights are on according to santas plan\n", onLights);
    printf("%d is the brightness according to santas plan", brightness);

    float seconds = (float)(end_time - start_time) / CLOCKS_PER_SEC;
    printf("\nNeeded Time: %f seconds", seconds);

    return 0;
}